include/seqan3/alphabet/concept.hpp

```diff
@@ -23,20 +21,67 @@
 #include <seqan3/core/type_traits/function.hpp>
 #include <seqan3/std/concepts>
 
+// ============================================================================
+// forwards
+// ============================================================================
+
+namespace seqan3::custom
+{
+
+/*!\brief A type that can be specialised to provide customisation point implementations so that third party types
+ *        model alphabet concepts.
+ * \tparam t The type you wish to specialise for.
+ * \ingroup alphabet
+ *
+ * \details
+ *
+ * For examples of when and how you can make use of this type, please see \link about_customisation the page on
+ * customisation \endlink and the \link howto_write_an_alphabet_custom section on third party types \endlink in
+ * the Alphabet HowTo.
+ *
+ * Please note that by default the `t const`, `t &` and `t const &` specialisations of this class inherit the
+ * specialisation for `t` so you usually only need to provide a specialisation for `t`.
+ *
+ * \note Only use this, if you cannot provide respective functions in your namespace.
+ */
+template <typename t>
+struct alphabet
+{};
+
+//!\cond
+template <typename t>
+struct alphabet<t const> : alphabet<t>
+{};
+
+template <typename t>
+struct alphabet<t &> : alphabet<t>
+{};
+
+template <typename t>
+struct alphabet<t const &> : alphabet<t>
+{};
+//!\endcond
+
+} // namespace seqan3::custom
+
 // ============================================================================
 // to_rank()
 // ============================================================================
 
-namespace seqan3::detail::adl::only
+namespace seqan3::detail::adl_only
 {
 
+//!\brief Poison-pill overload to prevent non-ADL forms of unqualified lookup.
+template <typename ...args_t>
+void to_rank(args_t ...) = delete;
+
 //!\brief Functor definition for seqan3::to_rank.
 struct to_rank_fn
 {
 private:
-    SEQAN3_CPO_IMPL(2, to_rank(v)                       )    // ADL
-    SEQAN3_CPO_IMPL(1, seqan3::custom::to_rank(v)       )    // customisation namespace
-    SEQAN3_CPO_IMPL(0, v.to_rank()                      )    // member
+    SEQAN3_CPO_IMPL(2, seqan3::custom::alphabet<decltype(v)>::to_rank(v))    // explicit customisation
```

Comment:
Just a remark. We want the user to prefer overload in the respective objects namespace if possible. But why do we ask now in this order? Would the compiler not stop when it knows it found a valid overload and now it always has to begin with the least likely scenario and therefor affect the compile time? Just a thought.

Reply:
I am pretty sure the compiler needs to instantiate all IMPLs to find out what the valid overload-set is so it doesn't make a difference for compile-time.

However, it might make sense to document more clearly that we want people to provide implementations in the reverse order if possible.